codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt

```diff
@@ -159,55 +153,115 @@ private class ServerHttpProtocolImplGenerator(
                     #{SmithyHttpServer}::protocols::check_json_content_type(req)?;
                     Ok($inputName(#{parse_request}(req).await?))
                 }
-            }""",
+            }""".trimIndent(),
             *codegenScope,
             "I" to inputSymbol,
             "parse_request" to serverParseRequest(operationShape)
         )
+
         // Implement Axum `IntoResponse` for non streaming output types.
         val outputName = "${operationName}${ServerHttpProtocolGenerator.OPERATION_OUTPUT_WRAPPER_SUFFIX}"
-        rustTemplate(
-            """
-            pub struct $outputName(#{O});
-            ##[#{Axum}::async_trait]
-            impl #{Axum}::response::IntoResponse for $outputName {
-                type Body = #{SmithyHttpServer}::Body;
-                type BodyError = <Self::Body as #{SmithyHttpServer}::HttpBody>::Error;
-
-                fn into_response(self) -> #{http}::Response<Self::Body> {
-                    match #{serialize_response}(&self.0) {
-                        Ok(response) => response,
-                        Err(e) => #{http}::Response::builder().body(Self::Body::from(e.to_string())).expect("unable to build response from error")
-                    }
-                }
-            }""",
-            *codegenScope,
-            "O" to outputSymbol,
-            "serialize_response" to serverSerializeResponse(operationShape)
-        )
+        val errorSymbol = operationShape.errorSymbol(symbolProvider)
+
+        val handleSerializeOutput = """
+            Ok(response) => response,
+            Err(e) => #{http}::Response::builder().body(Self::Body::from(e.to_string())).expect("unable to build response from output")
+        """.trimIndent()
         if (operationShape.errors.isNotEmpty()) {
-            // Implement Axum `IntoResponse` for non streaming error types.
-            val errorName = "${operationName}${ServerHttpProtocolGenerator.OPERATION_ERROR_WRAPPER_SUFFIX}"
             rustTemplate(
```

Comment:
Nit: While it is clear to me, I'd rather add some documentation here and in the else about the differences in generated code between fallible and non-fallible operations.